### ISSUE TITLE
Subtract storage cost from bond when returning

### ISSUE THREAD

**Issue by mikedotexe** [near-daos/sputnik-dao-contract#47]
This issue is in regards to the SputnikDAO v2 contract.

Since proposals are not removed when completed, (accepted, rejected…) they take up storage space. This will eventually chip away at the storage on the DAO contract. On a medium-to-long timeframe, this can suddenly make certain DAOs inoperable until someone funds DAO with more NEAR. This is preventable by taking into consideration the amount of storage a proposal uses.

For example, James Waugh set up an AstroDAO for Portland, Oregon. When you go through the interface of Astro, it asks you to start a DAO with 5 Ⓝ, and this is how much we have in that DAO.

Here is a transaction where a person creates a proposal to a DAO:
https://explorer.testnet.near.org/transactions/4i8XGzg8g3TLiZVQPyy8MLjB7RrMJnWKs3aKDkTjq8Gc
This costs the bond amount (1 Ⓝ) plus the transaction fee of:
0.0007097988201294 Ⓝ

Let's pretend that this DAO explicitly does function calls, member addition/removal, and payouts in fungible tokens.

After the above proposal is accepted or rejected, the bond is returned to the proposer, but the storage of the proposal persists. If you compare before and after the proposal, you'll see that the DAO contract loses:
0.000665812311892 Ⓝ

This is a low amount but means after a few thousand proposals the DAO will run out of the default 5 Ⓝ that AstroDAO asked for in the creation wizard process. Then no more proposals can be added until someone figures out they need to add more Ⓝ.

**Solution**
For the various proposal types, come up with expected or worst-case-scenario storage fees, and subtract that from the bond when it's being returned. This is not web2; not everything is free.

**Comment by swfsql** [comment]
I believe that issue #41 can be related to this as far as storage space cleanup is concerned.  

Given the results of the test that you made, the DAO is put in a "unstable" storage usage regarding their users. Should it always be "stable" towards it's users?  
What I mean by saying "stable toward the users", would be that all users could quit while asking for a total internal cleanup to get back their storage deposits, and the DAO would be able to return to exactly it's initial storage requirements, while returning everyone's funds.  
Another way of considering is to say that for being unstable, the DAO tends to lose some storage funds over time while it's being used, as pointed out by you - that would be some "negative" instability towards zero funds. Another option would be a positive instability towards (slowly) infinite funds - which can always be re-invested or shared among users at particular levels anyway.

If it were to be stable, various internal structures should relate themselves towards users accounts. For example, the situation in which an user is a member of various roles currently doesn't consumes storage attachment from that user specifically - if it's a member of zero or 10 roles doesn't change the storage requirements from that user (it does from the DAO as a whole), but should this be changed?  
This relates to many other internal aspects, such as how are votes/voters stored, and how are ongoing/done proposals stored as well.

**Comment by TrevorJTClarke** [comment]
@mikedotexe good call, will consider this in a more immediate context as some DAOs are starting to experience pain already.

I actually believe there are 2 outcomes needed:
1. Do the storage cost from bond like you mentioned (however this is a bummer for proposers)
2. Create a way for proposals to get reduced or removed from storage, post-finalization

For the second point, i will create another issue for discussion, as we potentially could alleviate DAO storage moving proposals to event standard if indexers can help UIs :D
